source/opt/insert_extract_elim.cpp

```diff
@@ -0,0 +1,125 @@
+static const int kSpvInsertObjectId = 0;
+static const int kSpvInsertCompositeId = 1;
+
+namespace spvtools {
+namespace opt {
+
+bool InsertExtractElimPass::ExtInsMatch(const ir::Instruction* extInst,
+    const ir::Instruction* insInst) const {
+  if (extInst->NumInOperands() != insInst->NumInOperands() - 1)
+    return false;
+  uint32_t numIdx = extInst->NumInOperands() - 1;
+  for (uint32_t i = 0; i < numIdx; ++i)
+    if (extInst->GetSingleWordInOperand(i + 1) !=
+        insInst->GetSingleWordInOperand(i + 2))
+      return false;
+  return true;
+}
+
+bool InsertExtractElimPass::ExtInsConflict(const ir::Instruction* extInst,
+    const ir::Instruction* insInst) const {
+  if (extInst->NumInOperands() == insInst->NumInOperands() - 1)
+    return false;
+  uint32_t extNumIdx = extInst->NumInOperands() - 1;
+  uint32_t insNumIdx = insInst->NumInOperands() - 2;
+  uint32_t numIdx = extNumIdx < insNumIdx ? extNumIdx : insNumIdx;
+  for (uint32_t i = 0; i < numIdx; ++i)
+    if (extInst->GetSingleWordInOperand(i + 1) !=
+        insInst->GetSingleWordInOperand(i + 2))
+      return false;
+  return true;
+}
+
+bool InsertExtractElimPass::EliminateInsertExtract(ir::Function* func) {
+  bool modified = false;
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+      switch (ii->opcode()) {
+      case SpvOpCompositeExtract: {
+        uint32_t cid = ii->GetSingleWordInOperand(kSpvExtractCompositeId);
+        ir::Instruction* cinst = def_use_mgr_->GetDef(cid);
+        uint32_t replId = 0;
+        while (cinst->opcode() == SpvOpCompositeInsert) {
+          if (ExtInsConflict(&*ii, cinst))
+            break;
+          if (ExtInsMatch(&*ii, cinst)) {
+            replId = cinst->GetSingleWordInOperand(kSpvInsertObjectId);
+            break;
+          }
+          cid = cinst->GetSingleWordInOperand(kSpvInsertCompositeId);
+          cinst = def_use_mgr_->GetDef(cid);
+        }
+        if (replId == 0)
+          break;
+        const uint32_t extId = ii->result_id();
+        (void)def_use_mgr_->ReplaceAllUsesWith(extId, replId);
+        def_use_mgr_->KillInst(&*ii);
+        modified = true;
+      } break;
+      default:
+        break;
+      }
+    }
+  }
+  return modified;
+}
+
+void InsertExtractElimPass::Initialize(ir::Module* module) {
+
+  module_ = module;
+
+  // Initialize function and block maps
+  id2function_.clear();
+  for (auto& fn : *module_)
+    id2function_[fn.result_id()] = &fn;
+
+  // Do def/use on whole module
+  def_use_mgr_.reset(new analysis::DefUseManager(consumer(), module_));
+};
+
+Pass::Status InsertExtractElimPass::ProcessImpl() {
+  bool modified = false;
+
+  // Process all entry point functions.
+  for (auto& e : module_->entry_points()) {
+    ir::Function* fn =
+        id2function_[e.GetSingleWordOperand(kSpvEntryPointFunctionId)];
+    modified = modified || EliminateInsertExtract(fn);
```

Comment:
Short circuiting seems the wrong thing here.    It means you'll stop after modifying only one entry point.   That will be the common case with glslang-generated modules, but wrong in general.

(This afflicts the other passes too, it seems.  I checked all three local_*cpp and inline_pass.cpp)